### PR TITLE
Introducing dbCheckpoint operator command

### PR DESCRIPTION
**Pull request by itsmeakashgoyal**
This PR includes changes to the operator command for creating db checkpoints/snapshots and retrieving information about all checkpoints that have been made.

### CODE REVIEW

File: bftengine/include/bftengine/DbCheckpointManager.hpp [hunk 1]
```diff
@@ -57,6 +56,19 @@ class DbCheckpointManager {
...
+  }
+
+  /***
+   * The operator command uses this function to find the next immediate stable seq number
+   * with checkpointWindowSize(150) where dbCheckpoint/snapshot will be created.
+   ***/
+  void setNextStableSeqNumToCreateSnapshot(const std::optional<SeqNum>& seqNum);
+  std::optional<SeqNum> getNextStableSeqNumToCreateSnapshot() const {
+    if (nextStableSeqNum_ == std::nullopt) return std::nullopt;
+    return nextStableSeqNum_;
```
Comment (nkumar04) [inline bftengine/include/bftengine/DbCheckpointManager.hpp:70]: nit: just return nextStableSeqNum_

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -4347,10 +4347,15 @@ ReplicaImp::ReplicaImp(bool firstTime,
...
                                        .count();
-      if (getReplicaConfig().dbCheckpointFeatureEnabled && seqNum && isCurrentPrimary() &&
-          !(seqNum % getReplicaConfig().dbCheckPointWindowSize) &&
-          (timeSinceLastSnapshot >= getReplicaConfig().dbSnapshotIntervalSeconds.count())) {
-        DbCheckpointManager::instance().sendInternalCreateDbCheckpointMsg(seqNum);
+      if ((getReplicaConfig().dbCheckpointFeatureEnabled && seqNum && isCurrentPrimary() &&
+           !(seqNum % getReplicaConfig().dbCheckPointWindowSize) &&
+           (timeSinceLastSnapshot >= getReplicaConfig().dbSnapshotIntervalSeconds.count())) ||
+          (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().has_value() &&
+           DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() == seqNum)) {
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.cpp:4354]: consider putting this condition in an inline function for better readability. Something like,
inline bool isOperatorInitiatedCmd(seqNum) {
   return (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().has_value() &&
           DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() == seqNum);
}

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 3]
```diff
@@ -4347,10 +4347,15 @@ ReplicaImp::ReplicaImp(bool firstTime,
...
+        DbCheckpointManager::instance().sendInternalCreateDbCheckpointMsg(seqNum, false);
+        if (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().has_value()) {
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.cpp:4356]: better to put this in an inline function. something like checkAndReset()

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 4]
```diff
@@ -5182,6 +5187,17 @@ void ReplicaImp::onExecutionFinish() {
     // We are about to stop execution. To avoid VC we now clear all pending requests
     clientsManager->clearAllPendingRequests();
   }
+
+  // Sending noop client requests to get the system into a stable checkpoint,
+  // allowing the create dbCheckpoint operator command to create a dbCheckpoint/snapshot.
+  if (getReplicaConfig().dbCheckpointFeatureEnabled && isCurrentPrimary() &&
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.cpp:5193]: Can we do this:

 if (getReplicaConfig().dbCheckpointFeatureEnabled && isCurrentPrimary() &&
      (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().has_value() &&
       DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() >= lastExecutedSeqNum)) {

       bool noop = (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() == lastExecutedSeqNum+1) ? false : true;    DbCheckpointManager::instance().sendInternalCreateDbCheckpointMsg(
        lastExecutedSeqNum + 1, noop);  // send noop till we reach stable checkpoint
       if(!noop) DbCheckpointManager::instance().setNextStableSeqNumToCreateSnapshot(std::nullopt);
       LOG_INFO(GL, "sendInternalCreateDbCheckpointMsg noop command ");
  }

With this, we do not need to modify the callback for creating DbCheckpoints.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 3 above)
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.cpp:4356]: we don't need this change. See the comment below

File: bftengine/include/bftengine/DbCheckpointManager.hpp (hunk 1 above)
Comment (itsmeakashgoyal) [reply bftengine/include/bftengine/DbCheckpointManager.hpp]: Fixed

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 4 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/ReplicaImp.cpp:5188]: Fixed

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 3 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Fixed

File: bftengine/include/bftengine/DbCheckpointManager.hpp [hunk 5]
```diff
@@ -110,7 +119,8 @@ class DbCheckpointManager {
   std::thread monitorThread_;
   std::mutex lock_;
   uint32_t maxNumOfCheckpoints_{0};  // 0-disabled
-  uint64_t lastCheckpointSeqNum_{0};
+  SeqNum lastCheckpointSeqNum_{0};
+  std::optional<SeqNum> nextStableSeqNum_ = std::nullopt;
```
Comment (nkumar04) [inline bftengine/include/bftengine/DbCheckpointManager.hpp:123]: nit: change name nextStableSeqNum_ -> nextSeqNumToCreateCheckPt_

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 6]
```diff
@@ -17,6 +17,7 @@
 #include <assertUtils.hpp>
 #include "Serializable.h"
 #include "db_checkpoint_msg.cmf.hpp"
+#include "concord.cmf.hpp"
```
Comment (nkumar04) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:20]: Do we need this here?

File: bftengine/src/bftengine/DbCheckpointManager.cpp [hunk 7]
```diff
@@ -263,4 +265,14 @@ void DbCheckpointManager::sendInternalCreateDbCheckpointMsg(const SeqNum& seqNum
...
\ No newline at end of file
+void DbCheckpointManager::setNextStableSeqNumToCreateSnapshot(const std::optional<SeqNum>& seqNum) {
+  if (seqNum == std::nullopt) {
+    nextStableSeqNum_ = std::nullopt;
+    return;
+  }
+  SeqNum seq_num_to_create_snapshot = (seqNum.value() + checkpointWindowSize);
+  seq_num_to_create_snapshot = seq_num_to_create_snapshot - (seq_num_to_create_snapshot % checkpointWindowSize);
+  nextStableSeqNum_ = seq_num_to_create_snapshot;
+  LOG_INFO(getLogger(), "setNextStableSeqNumToCreateSnapshot, nextStableSeqNum_: " << nextStableSeqNum_.value());
```
Comment (nkumar04) [inline bftengine/src/bftengine/DbCheckpointManager.cpp:276]: nit: you can use KVLOG

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 8]
```diff
@@ -5182,6 +5182,22 @@ void ReplicaImp::onExecutionFinish() {
     // We are about to stop execution. To avoid VC we now clear all pending requests
     clientsManager->clearAllPendingRequests();
   }
+
+  // Sending noop commands to get the system into a stable checkpoint,
+  // allowing the create dbCheckpoint operator command to create a dbCheckpoint/snapshot.
+  if (getReplicaConfig().dbCheckpointFeatureEnabled && isCurrentPrimary() &&
+      (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().has_value() &&
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.cpp:5189]: For better readability, can we wrap the condition in an inline function

```
bool DbCheckpointManager::isCreateDbCheckPtSeqNumSet(lastExecutedSeqNum) {
  return (DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().has_value() &&
       DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() >= lastExecutedSeqNum))
}
```

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 9]
```diff
@@ -192,6 +225,68 @@ async def test_db_checkpoint_cleanup(self, bft_network, tracker):
...
+        op = operator.Operator(bft_network.config, client, bft_network.builddir)
+        rep = await op.create_dbcheckpoint_cmd()
+        data = cmf_msgs.ReconfigurationResponse.deserialize(rep)[0]
+        assert(data.success == True)
+
+        """
+        dbcheckpoint will be created on next stable seq num 300.
+        Adding a 10-second delay to ensure that a dummy client request is sent and reaches a stable seq number of 300.
+        """
+        time.sleep(10)
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_dbsnapshot.py:250]: You can wait for next stable seqNum instead of sleep

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 10]
```diff
@@ -192,6 +225,68 @@ async def test_db_checkpoint_cleanup(self, bft_network, tracker):
...
+
+        getrep = await op.get_dbcheckpoint_info_request()
+        rsi_rep = client.get_rsi_replies()
+        data = cmf_msgs.ReconfigurationResponse.deserialize(getrep)[0]
+        assert(data.success == True)
+        for r in rsi_rep.values():
+            res = cmf_msgs.ReconfigurationResponse.deserialize(r)
+            assert(len(res[0].response.db_checkpoint_info) == 1)
+            dbcheckpoint_info_list = res[0].response.db_checkpoint_info
+            assert(any(dbcheckpoint_info.seq_num == 300 for dbcheckpoint_info in dbcheckpoint_info_list))
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_dbsnapshot.py:260]: Also you can check if db checkpoint is created using last block id from dbcheckpoint_info_list

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 11]
```diff
@@ -192,6 +225,68 @@ async def test_db_checkpoint_cleanup(self, bft_network, tracker):
...
+        rep = await op.get_dbcheckpoint_info_request()
+        rsi_rep = client.get_rsi_replies()
+        data = cmf_msgs.ReconfigurationResponse.deserialize(rep)[0]
+        assert(data.success == True)
+        for r in rsi_rep.values():
+            res = cmf_msgs.ReconfigurationResponse.deserialize(r)
+            assert(len(res[0].response.db_checkpoint_info) == 2)
+            dbcheckpoint_info_list = res[0].response.db_checkpoint_info
+            assert(any(dbcheckpoint_info.seq_num == 150 for dbcheckpoint_info in dbcheckpoint_info_list))
+            assert(any(dbcheckpoint_info.seq_num == 300 for dbcheckpoint_info in dbcheckpoint_info_list))
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_dbsnapshot.py:289]: same comment as above

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 12]
```diff
@@ -5182,6 +5182,22 @@ void ReplicaImp::onExecutionFinish() {
...
+       DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() >= lastExecutedSeqNum)) {
+    bool noop =
```
Comment (arc-vmware) [inline bftengine/src/bftengine/ReplicaImp.cpp:5191]: bool noop = DbCheckpointManager::instance().getNextStableSeqNumToCreateSnapshot().value() == (lastExecutedSeqNum + 1);

This is better, we donot need :? operator here.

File: bftengine/include/bftengine/DbCheckpointManager.hpp (hunk 5 above)
Comment (itsmeakashgoyal) [reply bftengine/include/bftengine/DbCheckpointManager.hpp]: Fixed

File: bftengine/src/bftengine/DbCheckpointManager.cpp (hunk 6 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/DbCheckpointManager.cpp]: Fixed

File: bftengine/src/bftengine/DbCheckpointManager.cpp (hunk 7 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/DbCheckpointManager.cpp]: Fixed

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 8 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/ReplicaImp.cpp]: Fixed

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 9 above)
Comment (itsmeakashgoyal) [reply tests/apollo/test_skvbc_dbsnapshot.py]: Fixed

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 10 above)
Comment (itsmeakashgoyal) [reply tests/apollo/test_skvbc_dbsnapshot.py:256]: Fixed. Verify that last dbCheckpoint is created with last block id 200.

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 11 above)
Comment (itsmeakashgoyal) [reply tests/apollo/test_skvbc_dbsnapshot.py:287]: Fixed. Verify that last dbCheckpoint is created with last block id 300.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 12 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/ReplicaImp.cpp:5189]: Fixed

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 13]
```diff
@@ -585,6 +586,11 @@ class ReplicaImp : public InternalReplicaApi, public ReplicaForStateTransfer {
 
   void setConflictDetectionBlockId(const ClientRequestMsg&, IRequestsHandler::ExecutionRequest&);
 
+  inline bool isCreateDbCheckPtSeqNumSet(const SeqNum& seqNum) {
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.hpp:589]: Can we move this to DbCheckPointManager. This api is not related to ReplicaImp class

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 14]
```diff
@@ -43,6 +43,7 @@
 #include "FakeClock.hpp"
 #include <ccron/ticks_generator.hpp>
 #include "EpochManager.hpp"
+#include "DbCheckpointManager.hpp"
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.hpp:46]: not reqruied, please see the comment below

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 15]
```diff
@@ -258,6 +254,8 @@ async def test_create_dbcheckpoint_cmd(self, bft_network, tracker):
             assert(len(res[0].response.db_checkpoint_info) == 1)
             dbcheckpoint_info_list = res[0].response.db_checkpoint_info
             assert(any(dbcheckpoint_info.seq_num == 300 for dbcheckpoint_info in dbcheckpoint_info_list))
+            # Verify that last dbCheckpoint is created with last block id 200.
+            assert(any(dbcheckpoint_info.block_id == 200 for dbcheckpoint_info in dbcheckpoint_info_list))
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_dbsnapshot.py:258]: What i meant is that you can call verify_snapshot_is_available(snapshot_id) to verify if checkpoint dir is created in the file system.

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 16]
```diff
@@ -287,6 +285,8 @@ async def test_get_dbcheckpoint_info_request_cmd(self, bft_network, tracker):
             dbcheckpoint_info_list = res[0].response.db_checkpoint_info
             assert(any(dbcheckpoint_info.seq_num == 150 for dbcheckpoint_info in dbcheckpoint_info_list))
             assert(any(dbcheckpoint_info.seq_num == 300 for dbcheckpoint_info in dbcheckpoint_info_list))
+            # Verify that last dbCheckpoint is created with last block id 300.
+            assert(any(dbcheckpoint_info.block_id == 300 for dbcheckpoint_info in dbcheckpoint_info_list))
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_dbsnapshot.py:289]: same as above

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 13 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/ReplicaImp.hpp]: Moved into DbCheckPointManager

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 14 above)
Comment (itsmeakashgoyal) [reply bftengine/src/bftengine/ReplicaImp.hpp]: removed

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 15 above)
Comment (itsmeakashgoyal) [reply tests/apollo/test_skvbc_dbsnapshot.py]: fixed.

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 16 above)
Comment (itsmeakashgoyal) [reply tests/apollo/test_skvbc_dbsnapshot.py]: fixed.

**Review by nkumar04 — APPROVED** [review]
LGTM!
Thanks!
